ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -2161,9 +2167,14 @@ private void getMaterializationMetadata(QB qb) throws SemanticException {
       return;
     }
     try {
-      gatherCTEReferences(qb, rootClause);
+      Map<String, CTEClause> materializationAliasToCTEs = new HashMap<>();
+      for (String key: this.aliasToCTEs.keySet()) {
+        CTEClause clause = this.aliasToCTEs.get(key);
+        materializationAliasToCTEs.put(key, new CTEClause(clause.alias, clause.cteNode, clause.withColList));
+      }
+      gatherCTEReferences(qb, rootClause, materializationAliasToCTEs);
       int threshold = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_CTE_MATERIALIZE_THRESHOLD);
-      for (CTEClause cte : Sets.newHashSet(aliasToCTEs.values())) {
+      for (CTEClause cte : Sets.newHashSet(materializationAliasToCTEs.values())) {
```

Comment:
I think the existing `CTEClause` objects stored in `this.aliasToCTEs` should be updated with the ones in `materializationAliasToCTEs`.
Some of the failing test passed when I added
```
this.aliasToCTEs.putAll(materializationAliasToCTEs);
```
because `gatherCTEReferences` collect the number of references of each CTE in the main query and sibling CTEs. The next few lines sets `cte.materialize` flag based on the number of references. But these changes are done in the instances in `materializationAliasToCTEs` only.
Later in `getMetadata` this flag is used to decide whether a CTE should be prematerialized or not and 
 `CTEClause` objects are pulled form `this.aliasToCTEs`
https://github.com/apache/hive/blob/de49826286e57976be447c00f9b9abfc41bc305e/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java#L2280
https://github.com/apache/hive/blob/de49826286e57976be447c00f9b9abfc41bc305e/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java#L2292

Reply:
I am adding 
> this.aliasToCTEs.putAll(materializationAliasToCTEs);

in `getMetaData` method. 

Adding it in this method (`getMaterializationMetadata`) doesn't work because then this.aliasToCTEs will have all CTE clauses. This is problematic because in getMetaData` method, we call an overloaded `getMetaData` method that tries to add the CTE clauses again into the map.